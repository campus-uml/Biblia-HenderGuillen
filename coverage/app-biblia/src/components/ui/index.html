
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for app-biblia/src/components/ui</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../../prettify.css" />
    <link rel="stylesheet" href="../../../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../../../index.html">All files</a> app-biblia/src/components/ui</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">67.09% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>312/465</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">83.33% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>35/42</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">50% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>11/22</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">67.09% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>312/465</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="Login.tsx"><a href="Login.tsx.html">Login.tsx</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="102" class="abs low">0/102</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="102" class="abs low">0/102</td>
	</tr>

<tr>
	<td class="file high" data-value="ReinaValeraBooks.tsx"><a href="ReinaValeraBooks.tsx.html">ReinaValeraBooks.tsx</a></td>
	<td data-value="93.45" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 93%"></div><div class="cover-empty" style="width: 7%"></div></div>
	</td>
	<td data-value="93.45" class="pct high">93.45%</td>
	<td data-value="168" class="abs high">157/168</td>
	<td data-value="81.48" class="pct high">81.48%</td>
	<td data-value="27" class="abs high">22/27</td>
	<td data-value="69.23" class="pct medium">69.23%</td>
	<td data-value="13" class="abs medium">9/13</td>
	<td data-value="93.45" class="pct high">93.45%</td>
	<td data-value="168" class="abs high">157/168</td>
	</tr>

<tr>
	<td class="file medium" data-value="api.tsx"><a href="api.tsx.html">api.tsx</a></td>
	<td data-value="50" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 50%"></div><div class="cover-empty" style="width: 50%"></div></div>
	</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="24" class="abs medium">12/24</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="24" class="abs medium">12/24</td>
	</tr>

<tr>
	<td class="file high" data-value="button.tsx"><a href="button.tsx.html">button.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="45" class="abs high">45/45</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="2" class="abs medium">1/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="45" class="abs high">45/45</td>
	</tr>

<tr>
	<td class="file medium" data-value="card.tsx"><a href="card.tsx.html">card.tsx</a></td>
	<td data-value="78.18" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 78%"></div><div class="cover-empty" style="width: 22%"></div></div>
	</td>
	<td data-value="78.18" class="pct medium">78.18%</td>
	<td data-value="55" class="abs medium">43/55</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="78.18" class="pct medium">78.18%</td>
	<td data-value="55" class="abs medium">43/55</td>
	</tr>

<tr>
	<td class="file medium" data-value="firebase.tsx"><a href="firebase.tsx.html">firebase.tsx</a></td>
	<td data-value="50" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 50%"></div><div class="cover-empty" style="width: 50%"></div></div>
	</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="32" class="abs medium">16/32</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="32" class="abs medium">16/32</td>
	</tr>

<tr>
	<td class="file high" data-value="header.tsx"><a href="header.tsx.html">header.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="39" class="abs high">39/39</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="39" class="abs high">39/39</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-02-15T19:41:38.557Z
            </div>
        <script src="../../../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../../../sorter.js"></script>
        <script src="../../../../block-navigation.js"></script>
    </body>
</html>
    