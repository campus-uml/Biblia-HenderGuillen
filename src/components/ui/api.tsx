import axios from 'axios';


const api = axios.create({
  baseURL: import.meta.env.VITE_BIBLE_API_URL,
  headers: {
    'api-key': import.meta.env.VITE_BIBLE_API_KEY,
  },
});


const reinaValeraId = import.meta.env.VITE_BIBLE_ID;


export const getReinaValeraBooks = async () => {
  const response = await api.get(`/bibles/${reinaValeraId}/books`);
  return response.data;
};

export const getBookChapters = async (bookId: string) => {
  const response = await api.get(`/bibles/${reinaValeraId}/books/${bookId}/chapters`);
  return response.data;
};

export const getChapterVerses = async (chapterId: string) => {
  const response = await api.get(`/bibles/${reinaValeraId}/chapters/${chapterId}/verses`);
  return response.data;
};

export const getVerseText = async (verseId: string) => {
  const response = await api.get(`/bibles/${reinaValeraId}/verses/${verseId}`);
  return response.data;
}