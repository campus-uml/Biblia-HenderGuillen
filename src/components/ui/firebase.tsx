import { initializeApp } from "firebase/app";
import { 
  getAuth, 
  GoogleAuthProvider, 
  GithubAuthProvider, 
  createUserWithEmailAndPassword, 
  signInWithEmailAndPassword 
} from "firebase/auth";

// Configuración de Firebase desde variables de entorno
const firebaseConfig = {
  apiKey: import.meta.env.VITE_FIREBASE_API_KEY, 
  authDomain: import.meta.env.VITE_FIREBASE_AUTH_DOMAIN,
  projectId: import.meta.env.VITE_FIREBASE_PROJECT_ID,
  storageBucket: import.meta.env.VITE_FIREBASE_STORAGE_BUCKET,
  messagingSenderId: import.meta.env.VITE_FIREBASE_MESSAGING_SENDER_ID,
  appId: import.meta.env.VITE_FIREBASE_APP_ID,
};

// Inicialización de Firebase
const app = initializeApp(firebaseConfig);

// Inicialización de servicios de Firebase
const auth = getAuth(app);

// Configuración de proveedores de autenticación
const googleProvider = new GoogleAuthProvider();
const githubProvider = new GithubAuthProvider();

// Función para registrar un usuario con correo y contraseña
const registerWithEmailAndPassword = async (email: string, password: string) => {
  try {
    const userCredential = await createUserWithEmailAndPassword(auth, email, password);
    return userCredential;
  } catch (error) {
    console.error("Error al registrar usuario:", error);
    throw error;
  }
};

// Función para iniciar sesión con correo y contraseña
const loginWithEmailAndPassword = async (email: string, password: string) => {
  try {
    const userCredential = await signInWithEmailAndPassword(auth, email, password);
    return userCredential;
  } catch (error) {
    console.error("Error al iniciar sesión:", error);
    throw error;
  }
};

export { 
  app, 
  auth, 
  googleProvider, 
  githubProvider, 
  registerWithEmailAndPassword, 
  loginWithEmailAndPassword 
};