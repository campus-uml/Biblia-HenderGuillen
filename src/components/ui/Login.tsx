import React, { useState } from "react";
import {
  loginWithEmailAndPassword,
  registerWithEmailAndPassword,
  googleProvider,
  githubProvider,
  auth,
} from "./firebase";
import { signInWithPopup } from "firebase/auth";
import { Button } from "./button";
import { Card, CardContent, CardHeader, CardTitle } from "./card";
import { FaGoogle, FaGithub, FaBook } from "react-icons/fa";

interface LoginProps {
  onLogin: (user: any) => void;
}

const Login: React.FC<LoginProps> = ({ onLogin }) => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState<string | null>(null);

  const handleLogin = async () => {
    try {
      const userCredential = await loginWithEmailAndPassword(email, password);
      console.log("Inicio de sesión exitoso");
      onLogin(userCredential.user);
    } catch (err: any) {
      setError(err.message);
    }
  };

  const handleRegister = async () => {
    try {
      const userCredential = await registerWithEmailAndPassword(email, password);
      console.log("Registro exitoso");
      onLogin(userCredential.user);
    } catch (err: any) {
      setError(err.message);
    }
  };

  const signInWithProvider = async (provider: any) => {
    try {
      const result = await signInWithPopup(auth, provider);
      console.log("Inicio de sesión con proveedor exitoso");
      onLogin(result.user);
    } catch (err: any) {
      console.error("Error al iniciar sesión con proveedor:", err);
      setError(err.message);
    }
  };

  return (
    <div className="min-h-screen flex items-center justify-center bg-gray-100">
      <Card className="w-full max-w-md min-h-[500px] flex flex-col justify-between bg-white p-6 rounded-lg shadow-xl">
        <CardHeader>
          <CardTitle className="text-center text-2xl font-bold">
            <div className="flex items-center justify-center space-x-2">
              <FaBook className="text-3xl text-yellow-600" />
              <span className="text-gray-700">Bienvenido a la Biblia Reina Valera</span>
            </div>
          </CardTitle>
        </CardHeader>
        <CardContent className="flex flex-col justify-center grow space-y-4">
          <div className="text-center text-lg text-gray-600 font-semibold">
            <p>Inicia sesión para continuar</p>
          </div>

          {error && (
            <div className="text-red-600 text-center bg-red-100 p-2 rounded-md">
              <p>{error}</p>
            </div>
          )}

          <input
            type="email"
            placeholder="Correo electrónico"
            value={email}
            onChange={(e) => setEmail(e.target.value)}
            className="w-full px-4 py-2 border rounded-lg focus:outline-none focus:ring-2 focus:ring-blue-500"
          />
          <input
            type="password"
            placeholder="Contraseña"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
            className="w-full px-4 py-2 border rounded-lg focus:outline-none focus:ring-2 focus:ring-blue-500"
          />

          <Button
            onClick={handleLogin}
            className="w-full bg-blue-600 text-white font-semibold py-2 rounded-lg hover:bg-blue-500 transition"
          >
            Iniciar sesión
          </Button>

          <Button
            onClick={handleRegister}
            className="w-full bg-green-600 text-white font-semibold py-2 rounded-lg hover:bg-green-500 transition"
          >
            Registrarse
          </Button>

          <div className="flex items-center gap-2 my-4">
            <hr className="flex-grow border-gray-300" />
            <span className="text-gray-500 font-medium whitespace-nowrap">O inicia sesión con</span>
            <hr className="flex-grow border-gray-300" />
          </div>

          <Button
            onClick={() => signInWithProvider(googleProvider)}
            className="w-full bg-red-600 text-white font-semibold py-2 rounded-lg hover:bg-red-500 flex items-center justify-center space-x-2 transition"
          >
            <FaGoogle className="text-xl" />
            <span>Ingresar con Google</span>
          </Button>

          <Button
            onClick={() => signInWithProvider(githubProvider)}
            className="w-full bg-gray-900 text-white font-semibold py-2 rounded-lg hover:bg-gray-700 flex items-center justify-center space-x-2 transition"
          >
            <FaGithub className="text-xl" />
            <span>Ingresar con GitHub</span>
          </Button>
        </CardContent>
      </Card>
    </div>
  );
};

export default Login;
