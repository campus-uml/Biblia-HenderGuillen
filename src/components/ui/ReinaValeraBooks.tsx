import React, { useEffect, useState } from 'react';
import { signOut } from 'firebase/auth';
import { auth } from './firebase'; // Aseg√∫rate de que esta ruta es correcta
import {
  getReinaValeraBooks,
  getBookChapters,
  getChapterVerses,
  getVerseText,
} from './api';
import { ArrowLeft } from 'lucide-react';

const ReinaValeraBooks: React.FC = () => {
  const [books, setBooks] = useState<any[]>([]);
  const [filteredBooks, setFilteredBooks] = useState<any[]>([]);
  const [selectedBook, setSelectedBook] = useState<any | null>(null);
  const [chapters, setChapters] = useState<any[]>([]);
  const [selectedChapter, setSelectedChapter] = useState<any | null>(null);
  const [verses, setVerses] = useState<any[]>([]);
  const [selectedVerse, setSelectedVerse] = useState<any | null>(null);
  const [verseText, setVerseText] = useState<string | null>(null);
  const [search, setSearch] = useState('');
  const [, setLoading] = useState(false);

  useEffect(() => {
    const fetchBooks = async () => {
      try {
        setLoading(true);
        const data = await getReinaValeraBooks();
        setBooks(data.data);
        setFilteredBooks(data.data);
      } catch (error) {
        console.error('Error fetching books:', error);
      } finally {
        setLoading(false);
      }
    };
    fetchBooks();
  }, []);

  useEffect(() => {
    setFilteredBooks(
      books.filter((book) =>
        book.name.toLowerCase().includes(search.toLowerCase())
      )
    );
  }, [search, books]);

  const handleBookClick = async (book: any) => {
    setSelectedBook(book);
    setSelectedChapter(null);
    setSelectedVerse(null);
    setVerseText(null);
    setVerses([]);
    try {
      setLoading(true);
      const data = await getBookChapters(book.id);
      setChapters(data.data);
    } catch (error) {
      console.error('Error fetching chapters:', error);
    } finally {
      setLoading(false);
    }
  };

  const handleChapterClick = async (chapter: any) => {
    setSelectedChapter(chapter);
    setSelectedVerse(null);
    setVerseText(null);
    try {
      setLoading(true);
      const data = await getChapterVerses(chapter.id);
      setVerses(data.data);
    } catch (error) {
      console.error('Error fetching verses:', error);
    } finally {
      setLoading(false);
    }
  };

  const handleVerseClick = async (verse: any) => {
    setSelectedVerse(verse);
    try {
      setLoading(true);
      const data = await getVerseText(verse.id);
      const parser = new DOMParser();
      const parsedDocument = parser.parseFromString(data.data.content, 'text/html');
      const cleanText = parsedDocument.body.textContent || '';
      setVerseText(cleanText);
    } catch (error) {
      console.error('Error fetching verse text:', error);
    } finally {
      setLoading(false);
    }
  };

  const handleLogout = async () => {
    try {
      await signOut(auth);
      window.location.reload(); // Recargar la p√°gina para redirigir al login
    } catch (error) {
      console.error("Error al cerrar sesi√≥n:", error);
    }
  };

  return (
    <div className="min-h-screen w-full bg-gradient-to-r from-purple-500 via-pink-500 to-red-500 flex flex-col items-center justify-center p-4">
      <div className="w-full max-w-6xl bg-white shadow-xl rounded-2xl p-6 flex flex-col items-center">
        <div className="w-full flex justify-between items-center mb-6">
          <h1 className="text-4xl font-bold text-gray-800">üìñ Reina Valera</h1>
          <button 
            className="bg-red-500 text-white px-4 py-2 rounded-lg shadow hover:bg-red-700 transition"
            onClick={handleLogout}
          >
            Cerrar sesi√≥n
          </button>
        </div>

        <input
          type="text"
          placeholder="Buscar libro..."
          className="p-4 border border-gray-300 rounded-lg w-full mb-6 focus:outline-none focus:ring-2 focus:ring-purple-500"
          value={search}
          onChange={(e) => setSearch(e.target.value)}
        />
 
        {!selectedBook ? (
          <div className="grid grid-cols-3 md:grid-cols-4 lg:grid-cols-6 gap-4 w-full">
            {filteredBooks.map((book) => (
              <div
                key={book.id}
                className="p-6 bg-purple-100 rounded-lg shadow-md cursor-pointer hover:bg-purple-200 hover:scale-105 transform transition-all duration-300 text-center"
                onClick={() => handleBookClick(book)}
              >
                <h2 className="text-lg font-bold text-purple-700">{book.name}</h2>
              </div>
            ))}
          </div>
        ) : !selectedChapter ? (
          <>
            <button
              className="mb-6 text-purple-600 flex items-center"
              onClick={() => setSelectedBook(null)}
            >
              <ArrowLeft className="mr-2" /> Volver
            </button>
            <div className="grid grid-cols-4 gap-4 w-full">
              {chapters.map((chapter) => (
                <div
                  key={chapter.id}
                  className="p-6 bg-pink-100 rounded-lg shadow-md cursor-pointer hover:bg-pink-200 hover:scale-105 transform transition-all duration-300 text-center"
                  onClick={() => handleChapterClick(chapter)}
                >
                  <h3 className="text-md font-medium text-pink-700">{chapter.reference}</h3>
                </div>
              ))}
            </div>
          </>
        ) : !selectedVerse ? (
          <>
            <button
              className="mb-6 text-pink-600 flex items-center"
              onClick={() => setSelectedChapter(null)}
            >
              <ArrowLeft className="mr-2" /> Volver
            </button>
            <div className="grid grid-cols-5 gap-4 w-full">
              {verses.map((verse) => (
                <div
                  key={verse.id}
                  className="p-6 bg-red-100 rounded-lg shadow-md cursor-pointer hover:bg-red-200 hover:scale-105 transform transition-all duration-300 text-center"
                  onClick={() => handleVerseClick(verse)}
                >
                  <p className="text-sm font-medium text-red-700">{verse.reference}</p>
                </div>
              ))}
            </div>
          </>
        ) : (
          <>
            <button
              className="mb-6 text-red-600 flex items-center"
              onClick={() => setSelectedVerse(null)}
            >
              <ArrowLeft className="mr-2" /> Volver
            </button>
            <p className="text-gray-800 bg-gray-100 p-6 rounded-lg shadow-inner whitespace-pre-wrap w-full text-center">{verseText || 'Cargando'}</p>
          </>
        )}
      </div>
    </div>
  );
};

export default ReinaValeraBooks;
