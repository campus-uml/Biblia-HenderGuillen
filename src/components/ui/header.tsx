import React, { useState, useEffect } from "react";
import { onAuthStateChanged, signOut } from "firebase/auth";
import { auth } from "./firebase";
import { FaSignOutAlt } from "react-icons/fa";

interface HeaderProps {
  onLogout: () => void;
}

const Header: React.FC<HeaderProps> = ({ onLogout }) => {
  const [user, setUser] = useState<any | null>(null);

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, (currentUser) => {
      setUser(currentUser);
    });
  
    // Asegúrate de que unsubscribe es una función antes de llamarla
    return () => {
      if (typeof unsubscribe === 'function') {
        unsubscribe();
      }
    };
  }, []);

  const handleLogout = async () => {
    await signOut(auth);
    onLogout();
  };

  return (
    <header className="bg-gradient-to-r from-red-300 to-purple-500 h-16 sm:h-20 px-4 sm:px-6 shadow-lg fixed top-0 left-0 w-full z-10">
      <div className="flex items-center justify-between h-full">
        <h1 className="text-lg sm:text-2xl font-bold text-white">Biblia Reina Valera</h1>
        {user ? (
          <div className="flex items-center space-x-4">
            <span className="text-white text-sm sm:text-base">{user.email}</span>
            <button
              onClick={handleLogout}
              className="p-2 rounded-full bg-red-500 text-white hover:bg-red-600 transition-all duration-300"
              aria-label="Cerrar sesión"
            >
              <FaSignOutAlt />
            </button>
          </div>
        ) : (
          <div className="text-white text-sm sm:text-base">Iniciar sesión</div>
        )}
      </div>
    </header>
  );
};

export default Header;