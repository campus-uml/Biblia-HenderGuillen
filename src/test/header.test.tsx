import { describe, it, expect, beforeEach, Mock, vi } from "vitest";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import Header from "../components/ui/header";
import '@testing-library/jest-dom';
import { auth } from "../components/ui/firebase";
import { onAuthStateChanged, signOut,  } from "firebase/auth";

// Mock de las funciones de Firebase
vi.mock("firebase/auth", () => ({
  getAuth: vi.fn(() => ({
    currentUser: null, // Simula que no hay usuario autenticado
  })),
  onAuthStateChanged: vi.fn(),
  signOut: vi.fn(),
  GoogleAuthProvider: vi.fn(),
  GithubAuthProvider: vi.fn(),
}));

describe("Testing de Header.tsx", () => {
  const mockOnLogout = vi.fn();

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("Debe renderizar el componente correctamente", () => {
    const { baseElement } = render(<Header onLogout={mockOnLogout} />);
    expect(baseElement).toMatchSnapshot();
  });

  it("Debe mostrar 'Iniciar sesión' cuando no hay un usuario autenticado", () => {
    (onAuthStateChanged as Mock).mockImplementation((_: any, callback: (arg0: null) => void) => {
      callback(null); // Simula que no hay usuario autenticado
      return vi.fn();
    });

    render(<Header onLogout={mockOnLogout} />);
    expect(screen.getByText(/Iniciar sesión/i)).toBeInTheDocument();
  });

  it("Debe mostrar el correo del usuario cuando está autenticado", async () => {
    const mockUser = { email: "test@example.com" };
    (onAuthStateChanged as Mock).mockImplementation((_: any, callback: (arg0: { email: string }) => void) => {
      callback(mockUser); // Simula un usuario autenticado
      return vi.fn();
    });

    render(<Header onLogout={mockOnLogout} />);
    expect(await screen.findByText(/test@example.com/i)).toBeInTheDocument();
  });

  it("Debe llamar a signOut y onLogout al hacer clic en el botón de logout", async () => {
    const mockUser = { email: "test@example.com" };
    (onAuthStateChanged as Mock).mockImplementation((_: any, callback: (arg0: { email: string }) => void) => {
      callback(mockUser); // Simula un usuario autenticado
      return vi.fn();
    });

    (signOut as Mock).mockResolvedValueOnce(undefined);

    render(<Header onLogout={mockOnLogout} />);
    const logoutButton = await screen.findByRole("button", { name: /Cerrar sesión/i });

    fireEvent.click(logoutButton);

    await waitFor(() => {
      expect(signOut).toHaveBeenCalledWith(auth); // Asegura que se llama a signOut
      expect(mockOnLogout).toHaveBeenCalled(); // Asegura que se llama a la función onLogout
    });
  });

  it("Debe limpiar la suscripción a onAuthStateChanged al desmontar el componente", () => {
    const unsubscribeMock = vi.fn();
    (onAuthStateChanged as Mock).mockImplementation(() => unsubscribeMock);

    const { unmount } = render(<Header onLogout={mockOnLogout} />);
    unmount();

    expect(unsubscribeMock).toHaveBeenCalled(); // Verifica que se llama a la función de desuscripción
  });
});