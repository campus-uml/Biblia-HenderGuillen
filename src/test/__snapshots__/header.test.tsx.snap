// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing de Header.tsx > Debe renderizar el componente correctamente 1`] = `
<body>
  <div>
    <header
      class="bg-gradient-to-r from-red-300 to-purple-500 h-16 sm:h-20 px-4 sm:px-6 shadow-lg fixed top-0 left-0 w-full z-10"
    >
      <div
        class="flex items-center justify-between h-full"
      >
        <h1
          class="text-lg sm:text-2xl font-bold text-white"
        >
          Biblia Reina Valera
        </h1>
        <div
          class="text-white text-sm sm:text-base"
        >
          Iniciar sesi√≥n
        </div>
      </div>
    </header>
  </div>
</body>
`;
